text/0066-separate-context-propagation.md

```diff
@@ -0,0 +1,606 @@
+with a program's application logic and cannot be cleanly encapsulated. All 
+concerns share an underlying distributed **context propagation** layer, for 
+storing state and accessing data across the lifespan of a distributed 
+transaction.
+
+
+# Cross-Cutting Concerns
+
+## Observability API
+Distributed tracing is one example of a cross-cutting concern. Tracing code is 
+interleaved with regular code, and ties together independent code modules which 
+would otherwise remain encapsulated. Tracing is also distributed, and requires 
+transaction-level context propagation in order to execute correctly.
+
+The various observability APIs are not described here directly. However, in this new 
+design, all observability APIs would be modified to make use of the generalized 
+context propagation mechanism described below, rather than the tracing-specific 
+propagation system it uses today.
+
+Note that OpenTelemetry APIs calls should *always* be given access to the entire 
+context object, and never just a subset of the context, such as the value in a 
+single key. This allows the SDK to make improvements and leverage additional 
+data that may be available, without changes to all of the call sites.
+
+## Correlations API
+
+In addition to trace propagation, OpenTelemetry provides a simple mechanism for 
+propagating metric labels, called the Correlations API. Correlations are 
+intended for indexing events in one service with attributes provided by a prior 
+service in the same transaction. This helps to establish a causal relationship 
+between these events. For example, determining that a particular browser version 
+is associated with a failure in an image processing service.
+
+Correlations are implemented as an extremely basic cross-cutting concern - a 
+transaction-level dictionary. All key-value pairs are stored in the context 
+object, and then injected by the Correlation propagator to be sent to the next 
+process. 
+
+The Correlations API is based on the [W3C Correlation-Context specification](https://w3c.github.io/correlation-context/), 
+and implements the protocol as it is defined in that working group.
+
+While Correlations can be used to prototype other cross-cutting concerns, this 
+mechanism is primarily intended to convey values for the OpenTelemetry 
+observability systems. New concerns with different criteria should be modeled 
+separately, using the same underlying context propagation layer as building 
+blocks.
+
+For backwards compatibility, OpenTracing Baggage is propagated as Correlations 
+when using the OpenTracing bridge.
+
+As every correlation adds additional overhead to every request, correlations 
+should be used with care. The implementation MUST propagate these values. It is 
+entirely on the user to choose when to drop correlations, either by removing 
+them via the API, or by filtering them via a network proxy or service mesh.
+
+**`GetCorrelation(context, key) -> value`**  
```

Comment:
I would remove completely the API definition for correlation from here. This is a sub-component of the context and we will discuss later when we define the specs for the correlation context.

Reply:
Ok, how about I add some langauge to say that this is just an example API? I actually tried to remove it entirely before, but then no one could understand the spec because there is no example usage for context propagation.

Reply:
You can say that it is an example (skeleton API) and remove a lot of the comments clarifications. We will add them in the specs very detailed.

Reply:
Done